### PR TITLE
WIP: ⬆️ Upgrade Director v2 service (Pydantic v2)

### PR DESCRIPTION
**Pull request by giancarloromeo**
## What do these changes do?
Upgrade Director-v2 service to Pydantic v2.

## Dev-ops checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

### CODE REVIEW

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_scheduler.py [hunk 1]
```diff
@@ -525,9 +529,9 @@ async def test_mark_all_services_in_wallet_for_removal(
         wallet_id = scheduler_data.wallet_info.wallet_id
         can_remove = scheduler_data.dynamic_sidecar.service_removal_state.can_remove
         match wallet_id:
-            case WalletID(1):
```
Comment (giancarloromeo) [inline services/director-v2/tests/unit/test_modules_dynamic_sidecar_scheduler.py:528]: match wallet_id:
>               case WalletID(1):
E               TypeError: called match pattern must be a type

File: packages/models-library/src/models_library/generated_models/docker_rest_api.py
Comment (giancarloromeo) [inline packages/models-library/src/models_library/generated_models/docker_rest_api.py:1]: Generated using datamodel-codegen tool.

File: packages/models-library/src/models_library/errors.py [hunk 2]
```diff
@@ -1,4 +1,6 @@
-from typing import Any, TypedDict
+from typing import Any
+from typing_extensions import TypedDict
```
Comment (giancarloromeo) [inline packages/models-library/src/models_library/errors.py:2]: https://docs.pydantic.dev/2.3/usage/types/dicts_mapping/#typeddict

File: services/director-v2/src/simcore_service_director_v2/core/settings.py [hunk 3]
```diff
@@ -56,11 +58,13 @@ class DirectorV0Settings(BaseCustomSettings):
...
-            scheme="http",
-            host=self.DIRECTOR_HOST,
-            port=f"{self.DIRECTOR_PORT}",
-            path=f"/{self.DIRECTOR_V0_VTAG}",
+        url: str = str(
+            AnyHttpUrl.build(
+                scheme="http",
+                host=self.DIRECTOR_HOST,
+                port=self.DIRECTOR_PORT,
+                path=f"{self.DIRECTOR_V0_VTAG}",
```
Comment (giancarloromeo) [inline services/director-v2/src/simcore_service_director_v2/core/settings.py:66]:
```shell
>>> from pydantic import AnyHttpUrl
>>> AnyHttpUrl.build(scheme="http", host="aaa", port=8080, path="/bbbbbbbb")
Url('http://aaa:8080//bbbbbbbb')
>>> AnyHttpUrl.build(scheme="http", host="aaa", port=8080, path="bbbbbbbb")
Url('http://aaa:8080/bbbbbbbb')
```

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_public.py [hunk 4]
```diff
@@ -81,7 +81,7 @@ def _dynamic_services_scheduler_settings(self) -> DynamicServicesSchedulerSettin
         return settings
 
     async def is_healthy(
-        self, dynamic_sidecar_endpoint: AnyHttpUrl, *, with_retry: bool = True
+        self, dynamic_sidecar_endpoint: str, *, with_retry: bool = True
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_public.py]: this is wrong, in the entire file. Please revert it.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_thin.py [hunk 5]
```diff
@@ -57,66 +56,62 @@ def __init__(self, app: FastAPI):
 
     def _get_url(
         self,
-        dynamic_sidecar_endpoint: AnyHttpUrl,
+        dynamic_sidecar_endpoint: str,
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_thin.py]: The same as in the above file. Please revert it everywhere in this file.

File: services/director-v2/src/simcore_service_director_v2/utils/osparc_variables.py [hunk 6]
```diff
@@ -89,7 +89,7 @@ def copy(
         return {k: self._variables_getters[k] for k in selection}
 
 
-_HANDLERS_TIMEOUT: Final[NonNegativeInt] = parse_obj_as(NonNegativeInt, 4)
+_HANDLERS_TIMEOUT: Final[NonNegativeInt] = TypeAdapter(NonNegativeInt).validate_python(4)
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/utils/osparc_variables.py:92]: maybe this one owls as well.
```suggestion
_HANDLERS_TIMEOUT: Final[NonNegativeInt] = 4
```

File: services/director-v2/tests/mocks/fake_task.json [hunk 7]
```diff
@@ -61,6 +61,7 @@
   "created": "1961-07-06T11:24:30.877Z",
   "modified": "2008-03-24T07:02:09.279Z",
   "last_heartbeat": null,
+  "pricing_info": null,
```
Comment (GitHK) [inline services/director-v2/tests/mocks/fake_task.json:64]: And now the the hidden issue here. What about existing data. Stored soemwhere in a DB maybe? Would this cause issues during the release if pricing_info is not automatically set as None.

@sanderegg you might have a better idea here?

**Review by GitHK — COMMENTED** [review]
please note a few places where things might be wrong

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_thin.py (hunk 5 above)
Comment (giancarloromeo) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_thin.py]: done

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_public.py (hunk 4 above)
Comment (giancarloromeo) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_public.py]: done

File: services/director-v2/src/simcore_service_director_v2/core/dynamic_services_settings/scheduler.py [hunk 8]
```diff
@@ -103,8 +103,8 @@ class DynamicServicesSchedulerSettings(BaseCustomSettings):
         ),
     )
 
-    DYNAMIC_SIDECAR_API_SAVE_RESTORE_STATE_TIMEOUT: PositiveFloat = Field(
-        60.0 * _MINUTE,
+    DYNAMIC_SIDECAR_API_SAVE_RESTORE_STATE_TIMEOUT: timedelta = Field(
```
[the remark this answers is not in the document]
Comment (giancarloromeo) [reply services/director-v2/src/simcore_service_director_v2/core/dynamic_services_settings/scheduler.py:106]:
```
.env-devel file:

[...]
DYNAMIC_SIDECAR_API_SAVE_RESTORE_STATE_TIMEOUT=01:00:00
[...]
```

File: services/director-v2/src/simcore_service_director_v2/models/comp_runs.py [hunk 9]
```diff
@@ -94,8 +94,8 @@ class Config:
                     "cluster_id": 0,
                     "iteration": 42,
                     "result": "NOT_STARTED",
-                    "created": "2021-03-01 13:07:34.19161",
-                    "modified": "2021-03-01 13:07:34.19161",
+                    "created": "2021-03-01T12:07:34.19161",
+                    "modified": "2021-03-01T13:07:34.19161",
```
[the remark this answers is not in the document]
Comment (giancarloromeo) [reply services/director-v2/src/simcore_service_director_v2/models/comp_runs.py:98]: checking..

File: services/director-v2/src/simcore_service_director_v2/models/comp_runs.py [hunk 10]
```diff
@@ -39,17 +39,17 @@ class CompRunsAtDB(BaseModel):
     run_id: PositiveInt
     project_uuid: ProjectID
     user_id: UserID
-    cluster_id: ClusterID | None
+    cluster_id: ClusterID | None = None
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/models/comp_runs.py:42]: no this is wrong, this comes from the DB and shall not have default values I think.

File: services/director-v2/src/simcore_service_director_v2/models/comp_runs.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/models/comp_runs.py]: what do you mean new format? this is the format given by sqlalchemy/postgres. how would the format change at all?

File: services/director-v2/src/simcore_service_director_v2/models/comp_runs.py [hunk 11]
```diff
@@ -105,10 +105,10 @@ class Config:
...
                     "iteration": 12,
                     "result": "SUCCESS",
-                    "created": "2021-03-01 13:07:34.19161",
-                    "modified": "2021-03-01 13:07:34.19161",
-                    "started": "2021-03-01 8:07:34.19161",
-                    "ended": "2021-03-01 13:07:34.10",
+                    "created": "2021-03-01T13:07:34.19161",
+                    "modified": "2021-03-01T14:07:34.19161",
+                    "started": "2021-03-01T08:07:34.19161",
+                    "ended": "2021-03-01T13:07:34.10",
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/models/comp_runs.py]: same here

File: services/director-v2/src/simcore_service_director_v2/models/comp_tasks.py [hunk 12]
```diff
@@ -105,15 +111,12 @@ class Config(ServiceOutput.Config):
...
-    class Config:
-        extra = Extra.forbid
-        orm_mode = True
+    model_config = ConfigDict(extra="forbid", from_attributes=True)
 
 
 class CompTaskAtDB(BaseModel):
     project_id: ProjectID
-    node_id: NodeID
+    node_id: Annotated[NodeID, PlainSerializer(str, return_type=str)]
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/models/comp_tasks.py:119]: what is this about?

File: services/director-v2/src/simcore_service_director_v2/models/comp_tasks.py [hunk 13]
```diff
@@ -223,20 +228,21 @@ class Config:
                         }
                     },
                     "image": image_example,
-                    "submit": "2021-03-01 13:07:34.19161",
+                    "submit": "2021-03-01T13:07:34.19161",
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/models/comp_tasks.py:231]: same question, why is there anyway a new format here?>

File: services/director-v2/src/simcore_service_director_v2/utils/computations.py [hunk 14]
```diff
@@ -123,18 +122,18 @@ async def find_deprecated_tasks(
...
     }
-    today = datetime.datetime.now(tz=datetime.timezone.utc)
+    today = dt.datetime.now(tz=dt.UTC)
 
     def _is_service_deprecated(service: dict[str, Any]) -> bool:
         if deprecation_date := service.get("deprecated"):
-            deprecation_date = parse_obj_as(
-                datetime.datetime, deprecation_date
-            ).replace(tzinfo=datetime.timezone.utc)
+            deprecation_date = dt.datetime.fromisoformat(deprecation_date).replace(
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/utils/computations.py:134]: just use `arrow.get(deprecation_date).datetime.replace(tzinfo=dt.UTC)' instead

File: services/director-v2/tests/mocks/fake_task.json (hunk 7 above)
Comment (sanderegg) [reply services/director-v2/tests/mocks/fake_task.json:64]: check the model of the DB. and check the values if it is not clear

File: services/director-v2/tests/unit/with_dbs/test_api_route_computations.py [hunk 15]
```diff
@@ -225,19 +229,30 @@ def mocked_catalog_service_fcts_deprecated(
...
+            }
+        ).model_dump(by_alias=True)
+
+        deprecated = {
+            "deprecated": (
+                dt.datetime.now(tz=dt.UTC) - dt.timedelta(days=1)
+            ).isoformat()
+        }
+
+        data = {**ServiceGet.model_config["json_schema_extra"]["examples"][0], **data_published, **deprecated}  # type: ignore
```
Comment (sanderegg) [inline services/director-v2/tests/unit/with_dbs/test_api_route_computations.py:248]: the type ignore that you are putting everywhere when using json_schema_extra can be fixed by using assertions.
this has the advantage of not ignoring potential pitfalls.

Comment (sanderegg) [reply services/director-v2/tests/unit/with_dbs/test_api_route_computations.py:248]: And 2nd question:
@pcrespov @giancarloromeo did you adjust the model of ServiceGet for the deprecation field? or where is it hidden now?

File: services/director-v2/src/simcore_service_director_v2/core/settings.py [hunk 16]
```diff
@@ -166,10 +164,12 @@ class AppSettings(BaseCustomSettings, MixinLoggingSettings):
...
     DIRECTOR_V2_PROFILING: bool = False
 
-    DIRECTOR_V2_REMOTE_DEBUGGING_PORT: PortInt | None
+    DIRECTOR_V2_REMOTE_DEBUGGING_PORT: PortInt | None = Field(default=None)
 
     # extras
-    SWARM_STACK_NAME: str = Field("undefined-please-check", env="SWARM_STACK_NAME")
+    SWARM_STACK_NAME: str = Field(
+        "undefined-please-check", validation_alias="SWARM_STACK_NAME"
+    )
```
[the remark this answers is not in the document]
Comment (giancarloromeo) [reply services/director-v2/src/simcore_service_director_v2/core/settings.py:172]: :eye: :eye:

File: packages/models-library/src/models_library/api_schemas_directorv2/clusters.py [hunk 17]
```diff
@@ -126,6 +125,7 @@ class ClusterCreate(BaseCluster):
                     "name": "My awesome cluster",
                     "type": ClusterTypeInModel.ON_PREMISE,
                     "endpoint": "https://registry.osparc-development.fake.dev",
+                    "owner": None,
```
[the remark this answers is not in the document]
Comment (giancarloromeo) [reply packages/models-library/src/models_library/api_schemas_directorv2/clusters.py:128]: Yup, the field was required before. Changed with a merge. I revert this

File: services/director-v2/src/simcore_service_director_v2/models/comp_runs.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (giancarloromeo) [reply services/director-v2/src/simcore_service_director_v2/models/comp_runs.py:42]: remaining from bump-pydantic execution

File: packages/models-library/src/models_library/aiodocker_api.py [hunk 18]
```diff
@@ -33,25 +34,22 @@ def convert_list_to_dict(cls, v):
 class AioDockerResources1(Resources1):
     # NOTE: The Docker REST API documentation is wrong!!!
     # Do not set that back to singular Reservation.
-    Reservation: ResourceObject | None = Field(
+    reservation: ResourceObject | None = Field(
         None, description="Define resources reservation.", alias="Reservations"
```
[the remark this answers is not in the document]
Comment (giancarloromeo) [reply packages/models-library/src/models_library/aiodocker_api.py:38]: See the note above

File: services/director-v2/tests/unit/with_dbs/test_api_route_computations.py (hunk 15 above)
Comment (giancarloromeo) [reply services/director-v2/tests/unit/with_dbs/test_api_route_computations.py:248]: The `deprecated` field was in the `ServiceGet` and the mock was wrong.

File: services/director-v2/tests/mocks/fake_task.json (hunk 7 above)
Comment (giancarloromeo) [reply services/director-v2/tests/mocks/fake_task.json:63]: `pricing_info` is a required-nullable field in the Model. This is a mock and the field was missing. I don't get the issue...

File: services/director-v2/src/simcore_service_director_v2/models/comp_tasks.py (hunk 12 above)
Comment (giancarloromeo) [reply services/director-v2/src/simcore_service_director_v2/models/comp_tasks.py]: Found a cleaner way.

File: packages/models-library/src/models_library/clusters.py [hunk 19]
```diff
@@ -221,14 +221,14 @@ class Cluster(BaseCluster):
...
 
-    @model_validator(mode="before")
+    @model_validator(mode="after")
     @classmethod
-    def check_owner_has_access_rights(cls, values):
-        is_default_cluster = bool(values["id"] == DEFAULT_CLUSTER_ID)
-        owner_gid = values["owner"]
+    def _check_owner_has_access_rights(cls, values):
+        is_default_cluster = bool(values.id == DEFAULT_CLUSTER_ID)
+        owner_gid = values.owner
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/models-library/src/models_library/clusters.py]: hmm. did you actually fix this or is there something else? we usually put at least a comment before resolving, just for the reviewer to know whether something was done or not.

File: services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_service_specs.py [hunk 20]
```diff
@@ -244,19 +251,19 @@ def expected_dynamic_sidecar_spec(
...
+                        '{"REGISTRY_AUTH":false,"REGISTRY_PATH":null,'
+                        '"REGISTRY_URL":"foo.bar.com","REGISTRY_USER":'
+                        '"test","REGISTRY_PW":"test","REGISTRY_SSL":false}'
                     ),
                     "DY_DOCKER_HUB_REGISTRY_SETTINGS": "null",
                     "DY_SIDECAR_AWS_S3_CLI_SETTINGS": (
-                        '{"AWS_S3_CLI_S3": {"S3_ACCESS_KEY": "12345678", "S3_BUCKET_NAME": "simcore", '
-                        '"S3_ENDPOINT": "http://172.17.0.1:9001", "S3_REGION": "us-east-1", "S3_SECRET_KEY": "12345678"}}'
+                        '{"AWS_S3_CLI_S3":{"S3_ACCESS_KEY":"12345678","S3_BUCKET_NAME":"simcore",'
+                        '"S3_ENDPOINT":"http://172.17.0.1:9001/","S3_REGION":"us-east-1","S3_SECRET_KEY":"12345678"}}'
```
[the remark this answers is not in the document]
Comment (giancarloromeo) [reply services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_service_specs.py:261]: We keep the / and will fix elsewhere.

File: services/director-v2/src/simcore_service_director_v2/cli/_core.py [hunk 21]
```diff
@@ -52,14 +52,14 @@ async def _initialized_app(only_db: bool = False) -> AsyncIterator[FastAPI]:
...
+def _get_dynamic_sidecar_endpoint(settings: AppSettings, node_id: NodeIDStr) -> str:
     dynamic_sidecar_names = DynamicSidecarNamesHelper.make(NodeID(node_id))
     hostname = dynamic_sidecar_names.service_name_dynamic_sidecar
     port = settings.DYNAMIC_SERVICES.DYNAMIC_SIDECAR.DYNAMIC_SIDECAR_PORT
-    url: AnyHttpUrl = parse_obj_as(AnyHttpUrl, f"http://{hostname}:{port}")  # NOSONAR
-    return url
+    url = AnyHttpUrl.build(  # pylint: disable=no-member
+        scheme="http", host=hostname, port=port
+    )
+    return f"{url}"
```
[the remark this answers is not in the document]
Comment (giancarloromeo) [reply services/director-v2/src/simcore_service_director_v2/cli/_core.py]: reverted

File: packages/models-library/src/models_library/clusters.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (giancarloromeo) [reply packages/models-library/src/models_library/clusters.py]: solved

**Review by GitHK — APPROVED** [review]
for me this can be merged as soon a director-v2 tests are green

**Comment by giancarloromeo** [comment]
@sanderegg, thank you for taking care after merge:

image
image

**Review by matusdrobuliak66 — APPROVED** [review]

**Pull request merged**